File: HW2/candy.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;
int recur_candy(vector<int> &arr, int my_candy, int index = 0)
{
    if (index == arr.size())
        return 0;
    if (arr[index] == my_candy)
        return 1 + recur_candy(arr, my_candy, index + 1);
    int a = recur_candy(arr, my_candy, index + 1);
    int b = recur_candy(arr, arr[index], index + 1) - 1;
    return max(a, b);
}
int recurDP_candy(vector<int> &arr, int my_candy, vector<vector<int>> &memo, int index = 0)
{
    if (index == arr.size())
        return 0;

    if (memo[index][my_candy] != INT_MIN)
        return memo[index][my_candy];
    if (arr[index] == my_candy)
        memo[index][my_candy] = 1 + recurDP_candy(arr, my_candy, memo, index + 1);
    else
        memo[index][my_candy] = max(recurDP_candy(arr, my_candy, memo, index + 1), recurDP_candy(arr, arr[index], memo, index + 1) - 1);

    return memo[index][my_candy];
}
int main()
{
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    vector<vector<int>> memo(n, vector<int>(4, INT_MIN));
    cout << recurDP_candy(arr, 1, memo) << endl;
}

File: HW2/candy.py
dp = []
n = int(input())
arr = list(map(int, input().split()))
for i in range(n+1):
    dp.append([0, 0, 0])
mycandy = [0, 1, 2]
for i in range(1, n+1):
    for j in range(3):
        if mycandy[j] == arr[i-1]:
            dp[i][j] = 1+dp[i-1][j]
        else:
            dp[i][j] = max(-1+dp[i-1][arr[i-1]], dp[i-1][j])
print(dp[n][2])
